### PR TITLE
[persist] Remove some unused code & cases in Persist's Spine

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

Refactoring, motivated by: #16607. In the short term, simplifying the structure of Spine means I have fewer cases to account for when roundtripping the structure. In the medium term, having fewer cases to account for in the Spine structure should make future work on Compaction 2.0 easier to reason about.

### Tips for reviewer

This is structured as a series of smallish changes, one per commit, in case it's easier to review that way also.

It was slightly trickier to remove the `None` case than I expected, even though we are not introducing structurally-empty batches, because it was also used as a placeholder value in some cases. Let me know if anything doesn't make sense!

`cargo bench` suggests that this is performance-neutral. I'll give nightlies a go at it as well.

I appreciate that this takes us away from the upstream Spine a bit. I think on balance it's worth it to make reasoning about the Persist case more straightforward, but I appreciate others may feel differently on this one!

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [inline src/persist-client/src/internal/trace.rs:273]: Dropped this mostly for expediency, but I'm happy to bring it back if we have plans for it!

Comment (danhhz) [reply src/persist-client/src/internal/trace.rs:273]: I found it quite useful while working on various compaction 2.0 branches, but since you'll be doing more of that now, happy to defer to you on the usefulness!

File: src/persist-client/src/internal/trace.rs
Comment (danhhz) [inline src/persist-client/src/internal/trace.rs:789]: my attempt at this has it mapping `None` to `Vacant` which I slightly prefer to the expect, but this is also fine if you feel strongly

```
                    if let Some(merged) = merged {
                        self.merging[position] = MergeState::Single(merged);
                    } else {
                        self.merging[position] = MergeState::Vacant;
                    }
```

File: src/persist-client/src/internal/trace.rs
Comment (danhhz) [inline src/persist-client/src/internal/trace.rs:988]: This commit to remove the Option from insert_at doesn't feel like an overall win to me, but maybe there's some context I'm missing?

File: src/persist-client/src/internal/trace.rs
Comment (danhhz) [inline src/persist-client/src/internal/trace.rs]: I stress a little about this early return diverging. Another approach would be to handle the two cases down in the `self.merging` paths below, which I slightly prefer. Thoughts?

Something like
```
            MergeState::Vacant => match batch {
                Some(batch) => self.merging[index] = MergeState::Single(batch),
                None => self.merging[index] = MergeState::Vacant,
            }
            MergeState::Single(old) => match batch {
                None => self.merging[index] = MergeState::Single(old),
                Some(batch) => {
                    let compaction_frontier = Some(self.since.borrow());
                    self.merging[index] = MergeState::begin_merge(old, batch, compaction_frontier);
                }
            }
```

File: src/persist-client/src/internal/trace.rs
Comment (danhhz) [inline src/persist-client/src/internal/trace.rs:1401]: Interesting! Not sure I remember enough of the context here to convince myself that it's safe to remove the behavior of `work` automatically converting the batch to a `Complete` when it's been fully fueled. Mind walking me through it?

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:789]: Don't feel strongly! (I think we could _probably_ eliminate this case entirely with a larger refactoring, but I was making some effort to avoid changing the API too drastically...)

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:988]: The motive is purely to not have to worry about stuff like [this comment](https://github.com/MaterializeInc/materialize/pull/26121#discussion_r1530787645).

In the old world, a `None` might represent either a structurally-empty batch or just not having a batch to rollup to a higher level in the rollup method. Since the first case has gone away, it seemed easier to handle the rollup case in the rollup method itself. (And it felt like a bit of a weird behaviour: as though `Vec::push` took an option and just ignored the `None` case, for example.)

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:1401]: Sure! Looking at: https://github.com/MaterializeInc/materialize/pull/26121/commits/3e94accdf761418bd31ee941d82a560f26d7f980#diff-be562d3f5bcccdd4feabf92e78906997d7addec7c9d7aacd446e8a2df0edbf26L967-L982
- We first apply work. This is the only place that work is applied, and that's the only place an in-progress merge could become complete.
- We then check if the merge is complete, and if so, extract the completed batch.

So a nontrivial `MergeVariant::Complete` was only ever very short-lived in practice: created on line 967 and destroyed on line 980. We do need to adjust the definition of `is_complete` (which is also only ever called here) but otherwise nothing else is affected by deferring the call to `done` a bit.

(Why mess with this at all? Removing the `MergeVariant::Complete(None)` case meant I couldn't keep the `std::mem::replace` above, and this was the simplest change I could find that made the ownership work.)

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:837]: Yeah... the only reason that we never store `None` batches in the Spine (that may need to be serialized etc.) is because we never call `exert`, so removing this unblocks the whole thing I think...

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:988]: I don't feel super passionate about this commit, so if that's not motivating I can skip it and find some other way to address your other comment!

Comment (danhhz) [reply src/persist-client/src/internal/trace.rs:989]: Our Vacant vs Single(None) slack convo enlightened me here. Keep it!

**Review by danhhz — APPROVED** [review]

**Comment by bkirwi** [comment]
I did a full nightly run... everything currently failing in nightly is flaky on main.

I'll wait until after the release cut before merging this in.

**Pull request merged**
